exchanges/gateio/gateio.go

```diff
@@ -424,169 +1012,2926 @@ func (g *Gateio) SendAuthenticatedHTTPRequest(ctx context.Context, ep exchange.U
+	params := url.Values{}
+	if subAccountUserID != "" {
+		params.Set("sub_uid", subAccountUserID)
+	}
+	var response []SubAccountCrossMarginInfo
+	return response, g.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, http.MethodGet, walletSubAccountCrossMarginBalances, params, nil, &response)
+}
+
+// GetSavedAddresses retrieves saved currency address info and related details.
+func (g *Gateio) GetSavedAddresses(ctx context.Context, currency currency.Code, chain string, limit uint64) ([]WalletSavedAddress, error) {
+	params := url.Values{}
+	if !currency.IsEmpty() {
+		params.Set("currency", currency.String())
+	}
+	if chain != "" {
+		params.Set("chain", chain)
+	}
+	if limit > 0 {
+		params.Set("limit", strconv.FormatUint(limit, 10))
+	}
+	var response []WalletSavedAddress
+	return response, g.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, http.MethodGet, walletSavedAddress, params, nil, &response)
+}
+
+// GetPersonalTradingFee retrieves personal trading fee
+func (g *Gateio) GetPersonalTradingFee(ctx context.Context, currencyPair currency.Pair) (*PersonalTradingFee, error) {
+	params := url.Values{}
+	if !currencyPair.Quote.IsEmpty() && !currencyPair.Base.IsEmpty() {
+		// specify a currency pair to retrieve precise fee rate
+		currencyPair.Delimiter = currency.UnderscoreDelimiter
+		params.Set("currency_pair", currencyPair.String())
+	}
+	var response PersonalTradingFee
+	return &response, g.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, http.MethodGet, walletTradingFee, params, nil, &response)
+}
+
+// GetUsersTotalBalance retrieves user's total balances
+func (g *Gateio) GetUsersTotalBalance(ctx context.Context, ccy currency.Code) (*UsersAllAccountBalance, error) {
+	params := url.Values{}
+	if !ccy.IsEmpty() {
+		params.Set("currency", ccy.String())
+	}
+	var response UsersAllAccountBalance
+	return &response, g.SendAuthenticatedHTTPRequest(ctx, exchange.RestSpot, http.MethodGet, walletTotalBalance, params, nil, &response)
+}
+
+// ********************************* Margin *******************************************
+
+// GetMarginSupportedCurrencyPairs retrieves margin supported currency pairs.
+func (g *Gateio) GetMarginSupportedCurrencyPairs(ctx context.Context) ([]MarginCurrencyPairInfo, error) {
+	var currenciePairsInfo []MarginCurrencyPairInfo
+	return currenciePairsInfo, g.SendHTTPRequest(ctx, exchange.RestSpot, marginCurrencyPairs, &currenciePairsInfo)
+}
+
+// GetMarginSupportedCurrencyPair retrieves margin supported currency pair detail given the currency pair.
+func (g *Gateio) GetMarginSupportedCurrencyPair(ctx context.Context, cp currency.Pair) (*MarginCurrencyPairInfo, error) {
+	if cp.IsEmpty() {
+		return nil, currency.ErrCurrencyPairEmpty
+	}
+	fPair, err := g.GetPairFormat(asset.Margin, true)
+	if err != nil {
+		fPair = currency.PairFormat{
+			Delimiter: currency.UnderscoreDelimiter,
+			Uppercase: true,
+		}
+	}
```

Comment:
This shouldn't fail and if it does across this entire PR I think would be more appropriate to return the error. So please change across the entire PR.